gateway/handler_defaults.go

```diff
@@ -0,0 +1,207 @@
+package gateway
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/textproto"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ipfs/go-libipfs/files"
+	mc "github.com/multiformats/go-multicodec"
+
+	ipath "github.com/ipfs/interface-go-ipfs-core/path"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+)
+
+func (i *handler) serveDefaults(ctx context.Context, w http.ResponseWriter, r *http.Request, maybeResolvedImPath ImmutablePath, immutableContentPath ImmutablePath, contentPath ipath.Path, begin time.Time, requestedContentType string, logger *zap.SugaredLogger) bool {
+	ctx, span := spanTrace(ctx, "ServeDefaults", trace.WithAttributes(attribute.String("path", contentPath.String())))
+	defer span.End()
+
+	var (
+		pathMetadata           ContentPathMetadata
+		bytesResponse          files.File
+		isDirectoryHeadRequest bool
+		directoryMetadata      *directoryMetadata
+		err                    error
+	)
+
+	switch r.Method {
+	case http.MethodHead:
+		var data files.Node
+		pathMetadata, data, err = i.api.Head(ctx, maybeResolvedImPath)
+		if !i.handleRequestErrors(w, contentPath, err) {
+			return false
+		}
+		defer data.Close()
+		if _, ok := data.(files.Directory); ok {
+			isDirectoryHeadRequest = true
+		} else if f, ok := data.(files.File); ok {
+			bytesResponse = f
+		} else {
+			webError(w, fmt.Errorf("unsupported response type"), http.StatusInternalServerError)
+			return false
+		}
+	case http.MethodGet:
+		// TODO: refactor below: we should not have 2x20 duplicated flow control when the only difference is ranges.
+		rangeHeader := r.Header.Get("Range")
+		if rangeHeader == "" {
```

Comment:
:thought_balloon: this is not a blocker, we can do that in future PRs, but the fact we duplicate flow control 1:1 hints at root for improvement.